gcs/constraints/smart_table.cc

```diff
@@ -0,0 +1,642 @@
+                            smart_entry_flags[binary_entry_data] = make_binary_entry_flag(initial_state, propagators, binary_entry.var_1, binary_entry.var_2, binary_entry.constraint_type);
+
+                        WeightedPseudoBooleanTerm t{1_i, smart_entry_flags[binary_entry_data]};
+                        WeightedPseudoBooleanTerm minus_t{-1_i, smart_entry_flags[binary_entry_data]};
+                        entry_flags_sum.emplace_back(t);
+                        entry_flags_neg_sum.emplace_back(minus_t);
+                    },
+                    [&](const UnarySetEntry & unary_set_entry) {
+                        auto var = unary_set_entry.var;
+                        WeightedPseudoBooleanTerms set_value_sum{};
+                        WeightedPseudoBooleanTerms neg_set_value_sum{};
+                        for (const auto & val : unary_set_entry.values) {
+                            WeightedPseudoBooleanTerm t{1_i, var == val};
+                            WeightedPseudoBooleanTerm minus_t{-1_i, var == val};
+                            set_value_sum.emplace_back(t);
+                            neg_set_value_sum.emplace_back(minus_t);
+                        }
+
+                        auto flag = unary_set_entry.constraint_type == IN ? propagators.create_proof_flag("inset") : propagators.create_proof_flag("notinset");
+
+                        propagators.define_pseudoboolean_ge(initial_state, move(set_value_sum), 1_i,
+                            unary_set_entry.constraint_type == IN ? flag : ! flag);
+                        propagators.define_pseudoboolean_ge(initial_state, move(neg_set_value_sum), 0_i,
+                            unary_set_entry.constraint_type == IN ? ! flag : flag);
+
+                        WeightedPseudoBooleanTerm t{1_i, flag};
+                        WeightedPseudoBooleanTerm minus_t{-1_i, flag};
+                        entry_flags_sum.emplace_back(t);
+                        entry_flags_neg_sum.emplace_back(minus_t);
+                    },
+                    [&](UnaryValueEntry unary_value_entry) {
+                        Literal l = literal_from_unary_entry(unary_value_entry);
+                        WeightedPseudoBooleanTerm t{1_i, l};
+                        WeightedPseudoBooleanTerm minus_t{-1_i, l};
+                        entry_flags_sum.emplace_back(t);
+                        entry_flags_neg_sum.emplace_back(minus_t);
+                    }}
+                    .visit(entry);
+            }
+            auto tuple_len = Integer{static_cast<long long>(entry_flags_sum.size())};
+            propagators.define_pseudoboolean_ge(initial_state, move(entry_flags_sum), tuple_len, pb_selectors[tuple_idx]);
+            propagators.define_pseudoboolean_ge(initial_state, move(entry_flags_neg_sum), -tuple_len + 1_i, ! pb_selectors[tuple_idx]);
+        }
+    }
+
+    // Trigger when any var changes? Is this overkill?
```

Comment:
I guess if variables only appear in, e.g. less-than tuples, they can get weaker triggers. Perhaps this is another argument in favour of calculating all variables ourselves rather than relying upon the user passing them in.